torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -483,21 +486,40 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       minEndTime = futureTimeouts_.begin()->first;
     }
 
+    auto shouldUpdateMinEndTimePredicate = [&, this]() -> bool {
+      // Notice, whoever modifying `rpcRunning_`
+      // must acquire lock on `futureMutex_`.
+      // Otherwise, this predicate could deadlock.
+      // If during evaluating the predicate, `::shutdown()` is called, then
+      // the predicate missed the notification before it started waiting
+      // on the cond var.
+      if (!rpcRunning_.load()) {
+        return true;
+      }
+      steady_clock_time_point minEndTimeInMap = kInfiniteTimeoutTimePoint;
+      if (futureTimeouts_.empty()) {
+        minEndTimeInMap = kInfiniteTimeoutTimePoint;
+      } else {
+        minEndTimeInMap = futureTimeouts_.begin()->first;
+      }
+      return minEndTimeInMap < minEndTime;
+    };
+
+    bool shouldUpdateMinEndTime = true;
     if (minEndTime == kInfiniteTimeoutTimePoint) {
-      futureTimeoutCV_.wait(lock);
+      futureTimeoutCV_.wait(lock, shouldUpdateMinEndTimePredicate);
     } else {
-      futureTimeoutCV_.wait_until(lock, minEndTime);
+      shouldUpdateMinEndTime = futureTimeoutCV_.wait_until(
+          lock, minEndTime, shouldUpdateMinEndTimePredicate);
     }
-
-    if (!rpcRunning_.load()) {
-      return;
+    if (shouldUpdateMinEndTime) {
+      continue;
```

Comment:
With this `continue`, I guess this means processing of the timed out futures is going to be skipped until we do another round of updating the `minEndTime`? If there is a timed out future, then there will be some delay in processing it, do we need to have this continue?

Reply:
@rohan-varma 

Sorry we thought we are closed on https://github.com/pytorch/pytorch/pull/30355#discussion_r349846014.

Two conditions it will stop waiting on the cond var.

1) The cond var is notified, it means there is an ealier `minEndTime` inserted into the map.
So it will go back to the beggining of the loop, update the `minEndTime` and 

    1) if `now >= the latest minEndTime`, it will **process the futures immediately** without waiting on the future.
    2) if `now < the latest minEndTime`, it will wai until `minEndTime`

2) Time is up. `now >= the latest minEndTime`. It will wake from the wait and and process the timed out futures.